### PR TITLE
8298482: Implement ParallelGC NUMAStats for Linux

### PR DESCRIPTION
**Pull request by nick-arm**
ParallelGC has a seemingly useful option -XX:+NUMAStats that prints detailed information in GC.heap_info about which NUMA node pages in the eden space are bound to.  However as far as I can tell this only ever worked on Solaris and is not implemented on any of the systems we currently support.  This patch implements it on Linux using the move_pages system call.

The function os::get_page_info() and accompanying struct page_info was just a thin wrapper around the Solaris meminfo(2) syscall and was never ported to other systems so I've just removed it rather than try to emulate its interface.

There's also a method MutableNUMASpace::LGRPSpace::scan_pages() which attempts to find pages on the wrong NUMA node and frees them so that they have another chance to be allocated on the correct node by the first-touching thread, but I think this has always been a no-op on non-Solaris so perhaps should also be removed.  On Linux it shouldn't be necessary as you can bind pages to the desired node directly.

I don't know what the performance of this option was like on Solaris but on Linux the move_pages call can be quite slow: I measured about 25ms/GB on my system.  At the moment we call LGRPSpace::accumulate_statistics() twice per GC cycle: I removed the second call as it's likely to see a lot of uncommitted pages if the spaces were just resized. MutableNUMASpace::print_on() also calls accumulate_statistics() directly and since that's the only place this data is used, maybe we can drop the call from MutableNUMASpace::accumulate_statistics() as well?

Example output:

```
 PSYoungGen      total 4290560K, used 835628K [0x00000006aac00000, 0x0000000800000000, 0x0000000800000000)
  eden space 3096576K, 1% used [0x00000006aac00000,0x00000007176a9f48,0x0000000767c00000)
    lgrp 0 space 1761280K, 2% used [0x00000006aac00000,0x00000006acfc4980,0x0000000716400000)
    local/remote/unbiased/uncommitted: 1671168K/0K/0K/90112K, large/small pages: 0/440320
    lgrp 1 space 1335296K, 46% used [0x0000000716400000,0x000000073c2abb18,0x0000000767c00000)
    local/remote/unbiased/uncommitted: 1335296K/0K/0K/0K, large/small pages: 0/333824
  from space 1193984K, 65% used [0x00000007b7200000,0x00000007e6b9c778,0x0000000800000000)
  to   space 1247232K, 0% used [0x0000000767c00000,0x0000000767c00000,0x00000007b3e00000)
```

After testing this with SPECjbb for a while I noticed some pages always end up bound to the wrong node.  I think this is a regression caused by JDK-8283935 but I'll raise a separate ticket for that.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298482](https://bugs.openjdk.org/browse/JDK-8298482): Implement ParallelGC NUMAStats for Linux

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**) ⚠️ Review applies to [c8333026](https://git.openjdk.org/jdk/pull/11635/files/c8333026f8641e99a1d8638ab4ce85b094f1e5c9)
 * [Stefan Johansson](https://openjdk.org/census#sjohanss) (@kstefanj - **Reviewer**) ⚠️ Review applies to [80bd78cd](https://git.openjdk.org/jdk/pull/11635/files/80bd78cd6c4d13919e71d0bbde65063606b6f0e6)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**) ⚠️ Review applies to [80bd78cd](https://git.openjdk.org/jdk/pull/11635/files/80bd78cd6c4d13919e71d0bbde65063606b6f0e6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11635/head:pull/11635` \
`$ git checkout pull/11635`

Update a local copy of the PR: \
`$ git checkout pull/11635` \
`$ git pull https://git.openjdk.org/jdk pull/11635/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11635`

View PR using the GUI difftool: \
`$ git pr show -t 11635`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11635.diff">https://git.openjdk.org/jdk/pull/11635.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/parallel/mutableNUMASpace.cpp [hunk 1]
```diff
@@ -892,28 +886,26 @@ void MutableNUMASpace::LGRPSpace::accumulate_statistics(size_t page_size) {
...
-        } else {
-          p += os::vm_page_size();
+  for (char *p = start; p < end; ) {
+    static const size_t PagesPerIteration = 128;
+    const void* pages[PagesPerIteration];
+    int lgrp_ids[PagesPerIteration];
+
+    size_t npages = 0;
+    for (; npages < PagesPerIteration && p < end; p += os::vm_page_size())
+      pages[npages++] = p;
```
Comment (albertnetymk) [inline src/hotspot/share/gc/parallel/mutableNUMASpace.cpp:896]: This means each page is `vm_page_size()` aligned. Would it be problematic if large-page is in use?

Comment (nick-arm) [reply src/hotspot/share/gc/parallel/mutableNUMASpace.cpp:896]: I don't think so: for HugeTLBFS it makes some redundant calls but in my testing always returns the node ID for the containing huge page even if the address given is not huge page aligned, and for THP you don't know whether a particular address is backed by a huge page so you have to step with the smallest granularity anyway. With large pages enabled the "small/large pages" count is misleading but I don't know any way to get this information on Linux except by parsing `/proc/self/smaps` which we probably don't want to do.

Comment (albertnetymk) [reply src/hotspot/share/gc/parallel/mutableNUMASpace.cpp:896]: I see; thanks for the clarification. Re the misleading "small/large pages" count, can it be dropped? Or, what useful info does this count convey towards identifying wrongly placed OS pages?

Comment (nick-arm) [reply src/hotspot/share/gc/parallel/mutableNUMASpace.cpp:893]: I don't think it adds any useful information now that the large page count is always zero: it's really just repeating the space size from the line above. I've removed it in the latest commit.

**Review by albertnetymk — APPROVED** [review]
> maybe we can drop the call from MutableNUMASpace::accumulate_statistics() as well?

Agree.

The rest look good.

File: src/hotspot/share/gc/parallel/mutableNUMASpace.cpp [hunk 2]
```diff
@@ -892,28 +877,23 @@ void MutableNUMASpace::LGRPSpace::accumulate_statistics(size_t page_size) {
...
+      for (size_t i = 0; i < npages; i++) {
+        if (lgrp_ids[i] < 0)
           space_stats()->_uncommited_space += os::vm_page_size();
-        }
-      } else {
-        return;
+        else if (lgrp_ids[i] == lgrp_id())
+          space_stats()->_local_space += os::vm_page_size();
+        else
+          space_stats()->_remote_space += os::vm_page_size();
```
Comment (kstefanj) [inline src/hotspot/share/gc/parallel/mutableNUMASpace.cpp:896]: Please add braces for the `for`-loop and the `if-else`-statements.

**Review by kstefanj — APPROVED** [review]
Looks good, just a style comment below.

Regarding the need of calls to `accumulate_statistics()` I also agree that the call in `print_on` should be enough.

**Review by tschatzl — APPROVED** [review]
Apart from the braces issue raised by Stefan, looks good.

File: src/hotspot/share/gc/parallel/mutableNUMASpace.cpp (hunk 2 above)
Comment (nick-arm) [reply src/hotspot/share/gc/parallel/mutableNUMASpace.cpp]: Done. Thanks for the reviews!

**Pull request closed without merging**